src/FSharpVSPowerTools.Logic/NavigableItemCache.fs

```diff
@@ -0,0 +1,139 @@
+namespace FSharpVSPowerTools.Navigation
+
+open System.ComponentModel.Composition
+open System
+open System.Collections.Concurrent
+open Microsoft.VisualStudio.Shell
+open EnvDTE80
+open Microsoft.VisualStudio.Shell.Interop
+open EnvDTE
+open FSharpVSPowerTools
+open System.Security.Cryptography
+open System.Text
+open Nessos.FsPickler
+open System.IO
+open System.Diagnostics
+open System.Threading
+open FSharpVSPowerTools.ProjectSystem
+
+type Source = string
+type FilePath = string
+
+type FileDescriptor =
+    { Path: FilePath
+      LastWriteTime: DateTime }
+
+type FileNavigableItems =
+    { Descriptor: FileDescriptor
+      Items: NavigableItem[] }
+
+[<Export>]
+type NavigableItemCache
+    [<ImportingConstructor>]
+    (
+        [<Import(typeof<SVsServiceProvider>)>] serviceProvider: System.IServiceProvider
+    ) =
+    
+    let cache = ConcurrentDictionary<FilePath, FileNavigableItems>()
+    let dirty = ref false
+    let pickler = FsPickler.CreateBinarySerializer()
+
+    let solutionPathHash (solutionPath: FilePath): string =
+        use digest = SHA1.Create()
+        let bytes = Encoding.UTF8.GetBytes solutionPath
+        digest.ComputeHash bytes |> Array.toHexString
+
+    let cacheFilePath solutionPath =
+        let solutionHash = solutionPathHash solutionPath
+        let root = Environment.ExpandEnvironmentVariables @"%LOCALAPPDATA%\VisualFSharpPowerTools\SolutionCaches"
+        root </> solutionHash </> "navigable_items.cache"
+
+    let loadFromDisk (solutionPath: FilePath) =
+        VSUtils.protect <| fun _ ->
+            cache.Clear()
+            let sw = Stopwatch.StartNew()
```

Comment:
Please use `Profiler` type from Core instead of `Stopwatch`. It will make it easier if we ever remove profiling code.